From 42adb869713d171253eb5afda3db10968551d175 Mon Sep 17 00:00:00 2001
From: Dan Scally <djrscally@gmail.com>
Date: Fri, 11 Sep 2020 14:35:34 +0100
Subject: [PATCH] Fix bugs in software_node_graph_get_next_endpoint()

As originally written, the software_node_graph_get_next_endpoint() function
holds on to references for port software_nodes and additionally passes invalid
combinations of parameters to software_node_get_next_child - an endpoint which
is not a child of the port that is also passed.

This patch rectifies both issues.

Signed-off-by: Dan Scally <djrscally@gmail.com>

(Patch from here:
- https://github.com/djrscally/miix-510-cameras/blob/18a7a4de9756235fab4c2328d946746ffe270afe/patches/0004-Fix-bugs-in-software_node_graph_get_next_endpoint.patch)
Signed-off-by: Tsuchiya Yuto <kitakar@gmail.com>
---
 drivers/base/swnode.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/base/swnode.c b/drivers/base/swnode.c
index 5cf9f1eef96f..80255e0b7739 100644
--- a/drivers/base/swnode.c
+++ b/drivers/base/swnode.c
@@ -563,6 +563,7 @@ software_node_graph_get_next_endpoint(const struct fwnode_handle *fwnode,
 {
 	struct swnode *swnode = to_swnode(fwnode);
 	struct fwnode_handle *old = endpoint;
+	struct fwnode_handle *parent_of_old;
 	struct fwnode_handle *parent;
 	struct fwnode_handle *port;
 
@@ -581,10 +582,22 @@ software_node_graph_get_next_endpoint(const struct fwnode_handle *fwnode,
 	}
 
 	for (; port; port = swnode_graph_find_next_port(parent, port)) {
+
+		if (old) {
+			parent_of_old = software_node_get_parent(old);
+
+			if (parent_of_old != port)
+				old = NULL;
+
+			fwnode_handle_put(parent_of_old);
+		}
+
 		endpoint = software_node_get_next_child(port, old);
 		fwnode_handle_put(old);
 		if (endpoint)
 			break;
+		else
+			fwnode_handle_put(port);
 	}
 
 	fwnode_handle_put(port);
-- 
2.25.1

